### PR TITLE
Feature/ev 204 calendar endpoint

### PR DESCRIPTION
**Pull request by turegjorup**

### CODE REVIEW

File: app/DoctrineMigrations/Version20190116141455.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        return '';
+    }
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf('mysql' !== $this->connection->getDatabasePlatform()->getName(), 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE itkdev_setting CHANGE name name VARCHAR(255) NOT NULL');
```
Comment (rimi-itk) [inline app/DoctrineMigrations/Version20190116141455.php:33]: Why this change?

File: src/AppBundle/Service/OccurrenceSplitterService.php [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+     * Create a collection of DailyOccurrences from an Occurrence.
+     *
+     * @param Occurrence $occurrence
+     * @param Collection $dailyOccurrences
+     *
+     * @throws \Exception
+     *
+     * @return Collection
+     */
+    private function createDailyOccurrenceCollection(Occurrence $occurrence, Collection $dailyOccurrences): Collection
```
Comment (rimi-itk) [inline src/AppBundle/Service/OccurrenceSplitterService.php]: Why are we passing a `Collection` to `createDailyOccurrenceCollection` (it seems to be always empty)? Why not let the function create the collection?

File: src/AppBundle/Service/OccurrenceSplitterService.php [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            while ($tempOccurrence->getStartDate() < $splitDate && $tempOccurrence->getEndDate() > $splitDate) {
+                $dailyOccurrence = $this->createDailyOccurrence($tempOccurrence->getStartDate(), $splitDate, $occurrence);
+                $dailyOccurrences->add($dailyOccurrence);
+
+                $tempOccurrence->setStartDate($splitDate);
+                $splitDate = $this->getFirstSplitDateTime($tempOccurrence->getStartDate());
+            }
+        }
+
+        $dailyOccurrence = $this->createDailyOccurrence($tempOccurrence->getStartDate(), $tempOccurrence->getEndDate(), $occurrence);
```
Comment (rimi-itk) [inline src/AppBundle/Service/OccurrenceSplitterService.php]: I'm not sure, but I have a suspicion that we may add an empty `DailyOccurrence` (with `startDate === endDate`) here in some edge cases. Maybe add a check for `$tempOccurrence->getStartDate() < $tempOccurrence->getEndDate()`?

File: src/AppBundle/Service/OccurrenceSplitterService.php [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+     * @throws \Exception
+     *
+     * @return Collection
+     */
+    private function createDailyOccurrenceCollection(Occurrence $occurrence, Collection $dailyOccurrences): Collection
+    {
+        $tempOccurrence = clone $occurrence;
+        $splitDate = $this->getFirstSplitDateTime($tempOccurrence->getStartDate());
+
+        if ($tempOccurrence->getStartDate() < $splitDate && $tempOccurrence->getEndDate() > $splitDate) {
```
Comment (rimi-itk) [inline src/AppBundle/Service/OccurrenceSplitterService.php]: I think we should add a limit on the number of `DailyOccurrence`s we allow to split into. If an event lasts for 87 years, say, this will run for a very long time.

We could either limit the number of `DailyOccurrence`s created, or skip splitting occurrences that lasts for more than a specific number of days.

File: src/AppBundle/Entity/DailyOccurrence.php [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+/**
+ * A daily occurrence of an Event.
+ *
+ * @ORM\Entity
+ *
+ * @ApiResource(
+ *   collectionOperations={
+ *     "get"={
+ *       "method"="GET",
+ *       "path"="/calendar",
```
Comment (rimi-itk) [inline src/AppBundle/Entity/DailyOccurrence.php]: Why not `/daily_occurrences`?

File: src/AppBundle/EventListener/OccurrenceListener.php [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+    /**
+     * Synchronize the DailyOccurrences to match the Occurrences scheduled for updates
+     * in doctrines unit of work.
+     *
+     * @param EntityManager $em
+     *
+     * @throws \Doctrine\ORM\ORMException
+     */
+    private function synchronizeDailyOccurrences(EntityManager $em): void
```
Comment (rimi-itk) [inline src/AppBundle/EventListener/OccurrenceListener.php:85]: Should this be called `updateDailyOccurrences` (to match `insert…` and `delete…`)?

File: src/AppBundle/EventListener/OccurrenceListener.php [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        foreach ($occurrences as $occurrence) {
+            $newDailyOccurrences = $this->occurrenceSplitter->getDailyOccurrences($occurrence);
+            $existingDailyOccurrences = $em->getRepository(DailyOccurrence::class)->findByOccurrence($occurrence);
+
+            // Loop through new DailyOccurrences and copy their data to the first exiting DailyOccurrence to update
+            // instead of doing delete/insert. Then remove the exiting DailyOccurrence.
+            // Once there are no more exiting DailyOccurrences we persist new entities.
+            $count = 0;
+            $totalExisting = \count($existingDailyOccurrences);
+            foreach ($newDailyOccurrences as $newDailyOccurrence) {
```
Comment (rimi-itk) [inline src/AppBundle/EventListener/OccurrenceListener.php:105]: Reusing occurrences is a good idea and the goal is to not change `id`s of existing occurrences (I suppose). However, unless we know (or assume) something about the order of occurrences we risk actually changing the `id`s anyway, e.g.

```
[
{ "startTime": '2001-01-01', 'endTime': '2002-01-01' },
{ "startTime": '2011-01-01', 'endTime': '2012-01-01' }
]
```

could become

```
[
{ "startTime": '2011-01-01', 'endTime': '2012-01-01' },
{ "startTime": '2001-01-01', 'endTime': '2002-01-01' }
]
```

This is a minor issue – and probably not a real issue at all …

**Review by rimi-itk — CHANGES_REQUESTED** [review]
@turegjorup LGTM. One could argue that the general clean up of configuration and migrations should not be a part of this pull request, but I don't want to make that argument.

I've added a few comment and questions.

File: app/DoctrineMigrations/Version20190116141455.php (hunk 1 above)
Comment (turegjorup) [reply app/DoctrineMigrations/Version20190116141455.php:33]: I suspect an update to a vendor dependency. The change is auto-generated but safe given that it's an increase in LENGTH

File: src/AppBundle/Service/OccurrenceSplitterService.php (hunk 2 above)
Comment (turegjorup) [reply src/AppBundle/Service/OccurrenceSplitterService.php]: Fixed - this was a leftover from the first implementation designed with recursion

File: src/AppBundle/Service/OccurrenceSplitterService.php (hunk 3 above)
Comment (turegjorup) [reply src/AppBundle/Service/OccurrenceSplitterService.php]: Fixed

File: src/AppBundle/Service/OccurrenceSplitterService.php (hunk 4 above)
Comment (turegjorup) [reply src/AppBundle/Service/OccurrenceSplitterService.php]: I agree this is not optimal but disagree with your solution. I think we should either:
* Leave as is
* Limit the allowed length of Occurrences to `< 1 year` or `< 10 years`
* Implement a job queue and handle splitting asynchronously.

I think having an implementation were `DailyOccurrences` match their respective `Occurrence` "except if..." is the wrong way to go on this.

File: src/AppBundle/Entity/DailyOccurrence.php (hunk 5 above)
Comment (turegjorup) [reply src/AppBundle/Entity/DailyOccurrence.php]: When we started talking about this it was `/calendar` - now changed to `/daily_occurrences` for consistency

File: src/AppBundle/EventListener/OccurrenceListener.php (hunk 6 above)
Comment (turegjorup) [reply src/AppBundle/EventListener/OccurrenceListener.php:85]: `synchronize` is a deliberate choice because this function both updates, inserts and deletes.

File: src/AppBundle/EventListener/OccurrenceListener.php (hunk 7 above)
Comment (turegjorup) [reply src/AppBundle/EventListener/OccurrenceListener.php:105]: We know that the order of `DailyOccurrences` is always chronological because they are always created and updated by our service. 

I considered a number of strategies for determining if a `DailyOccurrence` existed already. I rejected all of them because of complexity and decreased performance on large collections when doing the "exists in collection" check. 

The current implementation is not perfect. There are cases where it gives a sub-optimal result. Fx if the `startDate` of an `Occurrence` is moved forward by one day. Then all  `DailyOccurrences` will shift forward one day by id as opposed to just inserting one extra `DailyOccurrence`. But given the above considerations, and the fact that `/daily_occurrences` is not writable i settled on this compromise.
